lib/ansible/modules/cloud/amazon/ec2_vol.py

```diff
@@ -599,18 +684,21 @@ def main():
         # Check if there is a volume already mounted there.
         if device_name:
             if device_name in inst.block_device_mapping:
+                # Handle a case when instance and device_name is given instead of volume_id
+                volume = ec2.get_all_volumes(volume_ids=[inst.block_device_mapping[device_name].volume_id])[0]
+                changed = modify_volume(module, ec2, volume)
                 module.exit_json(msg="Volume mapping for %s already exists on instance %s" % (device_name, instance),
-                                 volume_id=inst.block_device_mapping[device_name].volume_id,
+                                 volume_id=volume.id,
                                  device=device_name,
-                                 changed=False)
+                                 changed=changed)
 
     # Delaying the checks until after the instance check allows us to get volume ids for existing volumes
     # without needing to pass an unused volume_size
     if not volume_size and not (id or name or snapshot):
         module.fail_json(msg="You must specify volume_size or identify an existing volume by id, name, or snapshot")
 
-    if volume_size and (id or snapshot):
-        module.fail_json(msg="Cannot specify volume_size together with id or snapshot")
+    if volume_size and id:
```

Comment:
snapshot and volume_size can be specified together? Did you mean to remove id, not snapshot? But can you change this to use mutually_exclusive while you're here? i.e. 
module = AnsibleModule(argument_spec=argument_spec,
                                          mutually_exclusive=[(volume_size, snapshot)])